src/unix.rs

```diff
@@ -7,27 +7,204 @@
+        siginfo.initialized = true;
+    }
+}
+
+struct EventedReceiver;
+
+impl Evented for EventedReceiver {
+    fn register(&self, poll: &MioPoll, token: Token, events: Ready, opts: PollOpt) -> io::Result<()> {
+        EventedFd(&GLOBALS.receiver).register(poll, token, events, opts)
+    }
+    fn reregister(&self, poll: &MioPoll, token: Token, events: Ready, opts: PollOpt) -> io::Result<()> {
+        EventedFd(&GLOBALS.receiver).reregister(poll, token, events, opts)
+    }
+    fn deregister(&self, poll: &MioPoll) -> io::Result<()> {
+        EventedFd(&GLOBALS.receiver).deregister(poll)
+    }
+}
+
+// There'll be stuff inside
+struct Driver {
+    id: CoreId,
+    wakeup: PollEvented<EventedReceiver>,
+}
+
+impl Future for Driver {
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<(), ()> {
+        // Drain the data from the pipe and maintain interest in getting more
+        let any_wakeup = self.drain();
+        if any_wakeup {
+            self.broadcast();
+        }
+        // This task just lives until the end of the event loop
+        Ok(Async::NotReady)
+    }
+}
 
+impl Drop for Driver {
+    fn drop(&mut self) {
+        let mut drivers = GLOBALS.drivers.lock().unwrap();
+        drivers.remove(&self.id);
+    }
+}
+
+impl Driver {
+    fn new(handle: &Handle) -> Self {
+        Driver {
+            id: handle.id(),
+            // TODO: Any chance of errors here?
+            wakeup: PollEvented::new(EventedReceiver, handle).unwrap(),
+        }
+    }
+    // Drain all data in the pipe and maintain an interest in read-ready
+    fn drain(&self) -> bool {
+        // Inform tokio we're interested in reading. It also hints on
+        // if we may be readable.
+        if let Async::NotReady = self.wakeup.poll_read() {
+            return false;
+        }
+        // Read all available data (until EAGAIN)
+        let mut received = false;
+        let mut buffer = [0; 1024];
+        loop {
+            match recv(GLOBALS.receiver, &mut buffer, MSG_DONTWAIT) {
+                Ok(0) => panic!("EOF on self-pipe"),
+                Ok(_) => received = true,
+                Err(NixError::Sys(Errno::EAGAIN)) => break,
+                Err(NixError::Sys(Errno::EINTR)) => (),
+                Err(e) => panic!("Bad read on self-pipe: {}", e),
+            }
+        }
+
+        // If we got here, it's because we got EAGAIN above. Ask for more data.
+        self.wakeup.need_read();
+
+        received
+    }
+    // Go through all the signals and broadcast everything
+    fn broadcast(&self) {
+        for (sig, value) in GLOBALS.pending.iter().enumerate() {
+            // Any signal of this kind arrived since we checked last?
+            if value.swap(false, Ordering::SeqCst) {
+                let signum = sig as c_int;
+                let mut siginfo = GLOBALS.signals[sig].lock().unwrap();
+                // It doesn't seem to be possible to do this through the iterators for now.
```

Comment:
I think it's ok to remove this comment, I got thrown for a spin a bit trying to figure out what "this" is.